File: 39-combination-sum/39-combination-sum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        res = []
        final = []
        n = len(candidates)
        def combine(i):
            if sum(res) == target:
                final.append(res[:])
                return
            if sum(res) > target or len(res) > 150:
                return
            for j in range(i,n):
                res.append(candidates[j])
                combine(j)
                res.pop()
        combine(0)
        return final
        
        

File: 39-combination-sum/39-combination-sum.java
class Solution {
    private List<List<Integer>> res = new LinkedList<>();
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        
        findSum(candidates, 0, 0, new LinkedList<>(), target);
        return res;
    }
    public void findSum(int[] candidates, int start, int tempSum, LinkedList<Integer> tempRes, int target){
        if(tempSum > target)return;
        if(tempSum == target){
            res.add((List) tempRes.clone());
            return;
        }
        for(int i = start; i < candidates.length; i++){
            int curVal = candidates[i];
            tempRes.addLast(curVal);
            findSum(candidates, i, tempSum + curVal, tempRes, target);
            tempRes.pollLast();
        }
    }
    
}